src/login/mod.rs

```diff
@@ -1,104 +1,220 @@
+
+        // Access token expired
+        // Refresh token before 20 seconds from actual expiration time to avoid minute details
+        if duration.num_seconds() >= -20 {
+            // Create oauth2 client
+            let client = BasicClient::new(
+                ClientId::new(CLIENT_ID.to_string()),
+                None,
+                AuthUrl::new(AUTH_URL.to_string()).expect("Invalid authorization endpoint URL"),
+                Some(TokenUrl::new(TOKEN_URL.to_string()).expect("Invalid token endpoint URL")),
+            )
+            .set_redirect_uri(
+                RedirectUrl::new(CALLBACK_URL.to_string()).expect("Invalid redirect URL"),
+            )
+            .set_auth_type(AuthType::RequestBody);
+
+            // Exchange refresh token with new access token
+            let refresh_token = user.get_refresh_token();
+            let token_response = client
+                .exchange_refresh_token(&RefreshToken::new(refresh_token.to_string()))
+                .request(http_client)
+                .expect("Failed to refresh OAuth access token");
+
+            // Set new access token
+            let access_token = token_response.access_token().secret();
+            user.set_oauth_token(access_token.to_string());
+
+            // Set new refresh token
+            let new_refresh_token = token_response.refresh_token();
+            if let Some(token) = new_refresh_token {
+                user.set_refresh_token(token.secret().to_string());
+            }
+
+            // Set new expiration time
+            let expires_in = token_response
+                .expires_in()
+                .expect("Failed to receive access expire time");
+            let expiration_time = Utc::now()
+                .checked_add_signed(Duration::from_std(expires_in).unwrap())
+                .unwrap()
+                .to_rfc3339();
+            user.set_expiration_time(expiration_time);
+
+            // Update configuration file on disk
+            let config_file = get_global_config_path();
+            let _ = user
+                .to_file(&config_file)
+                .expect("Failed to update configuration file");
         }
     }
-
-    anyhow::bail!(
-        "Timed out while waiting for API token. Try using `wrangler config` if login fails to work."
-    );
+    Ok(())
 }
 
-#[cfg(test)]
-mod tests {
-    use openssl::rsa::Rsa;
-
-    #[test]
-    fn test_rsa() {
-        let rsa = Rsa::generate(1024).unwrap();
-        rsa.public_key_to_pem_pkcs1().unwrap();
+// Invalidatess previous OAuth token if present
+fn invalidate_oauth_token() {
+    let user_create = GlobalUser::new();
+    if let Ok(user) = user_create {
+        if let GlobalUser::OAuthTokenAuth { .. } = user {
+            // Try to invalidate previous token
+            let result = revoke_token(&user);
+            if result.is_err() {
+                log::debug!("Failed to invalidate OAuth token before login.");
+            }
+        }
     }
 }
```

Comment:
Generally in this file `src/login/mod.rs` (besides tests), there is a lot of `unwrap` and `expect` usage. I think some of these are in places that will create a lot of noise from `wrangler report` usage, and could be cleaned up a bit. Consider some `?` usage in place of these where returning the error (even if it's printed later on) could give the user some way of debugging an issue. 

Many of them might be fine. Just adding a comment here to suggest taking another pass at error handling along this path.